src/main/java/edu/gatech/cs2340/trydent/animation/KeyframeAnimationBuilder.java

```diff
@@ -0,0 +1,303 @@
+        buildingFrame.duration = duration;
+        keyframes.add(buildingFrame);
+        Keyframe next = new Keyframe();
+        next.interpolation = buildingFrame.interpolation;
+        next.setPosition(buildingFrame.getPosition());
+        next.setRotation(buildingFrame.getRotation());
+        next.setScale(buildingFrame.getScale());
+        buildingFrame = next;
+        return this;
+    }
+
+    /**
+     * Adds a new keyframe of zero duration at the current position, rotation,
+     * and scale.
+     * <p>
+     * Unless this is the final keyframe in the animation, this call should be
+     * followed by a call to setTotalDuration(duration), which will
+     * automatically change the duration of this frame and other zero-duration
+     * frames to appropriately fill in the remaining time.
+     *
+     * @return
+     */
+    public KeyframeAnimationBuilder addKeyframe() {
+        return addKeyframe(0);
+    }
+
+    /**
+     * Sets the duration for the /TOTAL/ animation, changing the duration of all
+     * individual keyframes that have been set so far to accomplish this.
+     *
+     * If any keyframes have a negative or 0 duration, they will be used to fill
+     * in the gaps in time, and the other frames will be left alone.
+     *
+     * @param duration
+     * @return
+     */
+    public KeyframeAnimationBuilder setTotalDuration(double duration) {
+        if (duration <= 0) {
+            throw new TrydentException("The duration of the animation must be a positive number!");
+        }
+        enforcedTotal = duration;
+        return this;
+    }
+
+    /**
+     * If true, this animation will be constructed so that it loops back to the
+     * beginning when it ends.
+     *
+     * @param circular
+     * @return
+     */
+    public KeyframeAnimationBuilder setCircular(boolean circular) {
+        this.circular = circular;
+        return this;
+    }
+
+    /**
+     * Builds and returns a KeyframeAnimation with the given keyframes.
+     *
+     * @return
+     */
+    public KeyframeAnimation build() {
+        if (keyframes.isEmpty())
+            throw new TrydentException("Cannot build a keyframe animation without any frames!"
+                    + " Use addKeyframe(duration) to add a keyframe.");
+        enforceTotal();
+        double duration = 0;
+        int index = 0;
+        for (Keyframe frame : keyframes) {
+            if (++index == keyframes.size()) break;
+            duration += frame.duration;
+        }
+
+        if (keyframes.size() > 1) {
+            Keyframe last = keyframes.get(keyframes.size() - 1);
+            if (circular) {
+                last.duration = 1.0 / duration;
+            } else {
+                last.duration = 0;
+            }
+            duration += last.duration;
+        }
+
+        if (duration <= 0) {
+            throw new TrydentException("Total duration of animation must be positive!");
```

Comment:
Negative keyframes are fine on individual keyframes; they are treated like Keyframes of unspecified duration, which are filled in uniformly before the animation is built. E.g., if an animation consists of 10 keyframes of negative duration, and setTotalDuration(30.0) has been called, each keyframe will be changed to be 3 seconds long.